quiz/quiz.go

```diff
@@ -0,0 +1,117 @@
+package main
+
+import (
+	"bufio"
+	"encoding/csv"
+	"flag"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"time"
+	"unicode"
+)
+
+type problems struct {
+	q string
+	a string
+}
+
+func main() {
+	const defaultName = "problems.csv"
+	const defaultTimeLimit = 30
+	fileName := flag.String("csv", defaultName, "csv file in the format: `question,answer`")
+	timeLimit := flag.Int("limit", defaultTimeLimit, "Integer time limit ")
+	flag.Parse()
+
+	if isValidFileName(*fileName) {
+		arr, err := readFile(*fileName)
+		if err != nil {
+			log.Fatal(err)
+			return
+		}
+		problems := parseQuiz(arr)
+		quiz(problems, *timeLimit)
+	} else {
+		log.Fatal("Invalid file name")
+		return
+	}
+}
+
+func quiz(problems []problems, timeLimit int) {
+	correct := 0
+	incorrect := 0
+	timer := time.NewTimer(time.Duration(timeLimit) * time.Second)
+
+	for idx, prob := range problems {
+		fmt.Printf("Question number %d is %s? \n", idx+1, prob.q)
+		select {
+		case <-timer.C:
+			fmt.Println("Ops! You ran out of time")
+			fmt.Printf("out of %d questions, you got %d questions correct \n", len(problems), correct)
+			return
+		default:
+			var response string
+			fmt.Scanln(&response)
+			if response == prob.a {
+				fmt.Println("Genius! You got the answer right")
+				correct++
+			} else {
+				fmt.Println("You failed that question but that doesn't make you a failure :)")
+				incorrect++
+			}
+		}
+	}
+	fmt.Printf("out of %d questions, you got %d questions correct \n", len(problems), correct)
+	return
+}
+
+func parseQuiz(arr [][]string) []problems {
+	ret := make([]problems, len(arr))
+	for i, row := range arr {
+		ret[i] = problems{
+			q: row[0],
+			a: row[1],
+		}
+	}
+	return ret
+}
+
+func readFile(fileName string) ([][]string, error) {
+	problems, err := os.Open(fileName)
+	if err != nil {
+		fmt.Printf("Failed to open file with filename %s \n", fileName)
+		os.Exit(1)
+	}
+	header, err := bufio.NewReader(problems).ReadSlice('\n')
+	if err != nil {
+		return nil, err
+	}
+	_, err = problems.Seek(int64(len(header)), io.SeekStart)
+	if err != nil {
+		return nil, err
+	}
+	updatedProblems := csv.NewReader(problems)
+
+	rows, err := updatedProblems.ReadAll()
+	if err != nil {
+		return nil, err
+	}
+	return rows, err
+}
+
+func isValidFileName(fileName string) bool {
```

Comment:
Maybe this can be renamed to isValidCSVFile? because this is checking that it is a valid csv file.
Also, did you consider trying out os.Stat and os.IsNotExist to check that the file is valid and exists rather than the manual checks you did below? Or you can just check that the file extension is csv and then do os.Open like you did in line 81. Maybe this function is not really needed afterall?

Reply:
This only checks that the filename is valid. It doesn't actually check that the file itself is valid. I didn't know about the methods you talked about, I'll check them out.

Reply:
Sorry for being annoying :). According to some people on Reddit, `os.Open` does the work for me. Notice that I check for errors while trying to open the file. If an error exists, I return that error, if not, I move on with reading the file https://github.com/Del-sama/gophercises/blob/8a52970ca34334592cfebc0050ee3615ef901c60/quiz/quiz.go#L81

Reply:
What I was trying to explain was, rather than do all the steps you did, you could just check if it has a .csv extension, and proceed to os.Open, and os.Open will always return an error if the file name is not valid or it does not exist or anything like that.

Reply:
That's the thing. The user can create a file with an invalid file name. The file will exist but the filename will be wacky. os.Open won't throw an error then.